includes/acl/compression/impl/compact_constant_streams.h

```diff
@@ -149,10 +342,195 @@ namespace acl
+							if (!rtm::vector_all_near_equal3(adjusted_local_transform.translation, constant_translation, 0.0F))
+							{
+								any_constant_changed.translation = true;
+								constant_changed.translation = true;
+								adjusted_local_transform.translation = constant_translation;
+								raw_bone_stream.translations.set_raw_sample(sample_index, constant_translation);
+							}
+							ACL_ASSERT(bone_stream.translations.get_num_samples() == 1, "Constant translation stream mismatch!");
+							ACL_ASSERT(rtm::vector_all_near_equal3(bone_stream.translations.get_raw_sample<rtm::vector4f>(0), constant_translation, 0.0F), "Constant translation mismatch!");
+						}
+						if (has_scale && bone_stream.is_scale_constant)
+						{
+							const rtm::vector4f constant_scale = bone_range.scale.get_min();
+							if (!rtm::vector_all_near_equal3(adjusted_local_transform.scale, constant_scale, 0.0F))
+							{
+								any_constant_changed.scale = true;
+								constant_changed.scale = true;
+								adjusted_local_transform.scale = constant_scale;
+								raw_bone_stream.scales.set_raw_sample(sample_index, constant_scale);
+							}
+							ACL_ASSERT(bone_stream.scales.get_num_samples() == 1, "Constant scale stream mismatch!");
+							ACL_ASSERT(rtm::vector_all_near_equal3(bone_stream.scales.get_raw_sample<rtm::vector4f>(0), constant_scale, 0.0F), "Constant scale mismatch!");
+						}
+
+						rtm::qvvf& adjusted_object_transform = adjusted_object_pose[bone_index];
+						if (parent_bone_index == k_invalid_track_index)
+						{
+							adjusted_object_transform = adjusted_local_transform;	// Just copy the root as-is, it has no parent and thus local and object space transforms are equal
+						}
+						else
+						{
+							const DirtyState& parent_constant_changed = dirty_states[parent_bone_index];
+							const rtm::qvvf& parent_adjusted_object_transform = adjusted_object_pose[parent_bone_index];
+
+							if (bone_stream.is_rotation_constant && !constant_changed.rotation)
+							{
+								constant_changed.rotation = parent_constant_changed.rotation;
+							}
+							if (bone_stream.is_translation_constant && !constant_changed.translation)
+							{
+								constant_changed.translation = parent_constant_changed.translation;
+							}
+							if (has_scale && bone_stream.is_scale_constant && !constant_changed.scale)
+							{
+								constant_changed.scale = parent_constant_changed.scale;
+							}
+
+							// Compensate for the constant changes in your ancestors.
+							if (!bone_stream.is_rotation_constant && parent_constant_changed.rotation)
+							{
+								ACL_ASSERT(any_constant_changed.rotation, "No rotations have changed!");
+								adjusted_local_transform.rotation = rtm::quat_normalize(rtm::quat_mul(original_object_transform.rotation, rtm::quat_conjugate(parent_adjusted_object_transform.rotation)));
+								raw_bone_stream.rotations.set_raw_sample(sample_index, adjusted_local_transform.rotation);
```

Comment:
I'm not sure if we should modify the raw data. By doing so, any error we measure afterwards will no longer reflect what the true error is compared to the original raw clip. For example, a leaf bone might end up moved through error correction and even if it is constant, in local space, it would report an error of 0cm even though it no longer matches the original raw value.

But, if we don't modify the raw value accordingly, the bit rate optimization might try to reach a target that is unreachable... this might lead to more bits being used in vain.

Either way, we need to add a comment here to clarify why we are doing this as it is not obvious.

Reply:
Another fair point on the comment.

We aren't modifying raw data here.  We're modifying the raw channels generated from the raw data.  The raw data is left alone, and is still used at the end of the process to do regression testing.  Regression results are much tighter with this change than without.  My worst-case animation doesn't fail with a 2m error anymore(and doesn't glitch ingame anymore either).  It settles nicely near the 0.001 mark, which matches compression precision.  I suspect that regression tolerance can be reduced to compression precision now, as opposed to what it's been using.


>  bit rate optimization might try to reach a target that is unreachable... this might lead to more bits being used in vain.

Correct.  This is what was happening before ACL_COMPRESSION_OPTIMIZED.